docs/source/10_deployment/dask.md

```diff
@@ -0,0 +1,313 @@
+        """Factory method for creating the default dataset for the runner.
+
+        Args:
+            ds_name: Name of the missing dataset.
+
+        Returns:
+            An instance of ``_DaskDataSet`` to be used for all
+            unregistered datasets.
+        """
+        return _DaskDataSet(ds_name)
+
+    def _run_node(
+        self,
+        node: Node,
+        catalog: DataCatalog,
+        is_async: bool = False,
+        run_id: str = None,
+        *dependencies: Node,
+    ) -> Node:
+        """Run a single `Node` with inputs from and outputs to the `catalog`.
+
+        Wraps ``run_node`` to accept the set of ``Node``s that this node
+        depends on. When ``dependencies`` are futures, Dask ensures that
+        the upstream node futures are completed before running ``node``.
+
+        Args:
+            node: The ``Node`` to run.
+            catalog: A ``DataCatalog`` containing the node's inputs and outputs.
+            is_async: If True, the node inputs and outputs are loaded and saved
+                asynchronously with threads. Defaults to False.
+            run_id: The id of the pipeline run.
+            dependencies: The upstream ``Node``s to allow Dask to handle
+                dependency tracking. Their values are not actually used.
+
+        Returns:
+            The node argument.
+        """
+        return run_node(node, catalog, is_async, run_id)
+
+    def _run(
+        self, pipeline: Pipeline, catalog: DataCatalog, run_id: str = None
+    ) -> None:
+        nodes = pipeline.nodes
+        load_counts = Counter(chain.from_iterable(n.inputs for n in nodes))
+        node_dependencies = pipeline.node_dependencies
+        node_futures = {}
+
+        client = Client.current()
+        for node in nodes:
+            dependencies = (
+                node_futures[dependency] for dependency in node_dependencies[node]
+            )
+            node_futures[node] = client.submit(
+                self._run_node, node, catalog, self._is_async, run_id, *dependencies
+            )
+
+        for i, (_, node) in enumerate(
+            as_completed(node_futures.values(), with_results=True)
+        ):
+            self._logger.info("Completed node: %s", node.name)
+            self._logger.info("Completed %d out of %d tasks", i + 1, len(nodes))
+
+            # Decrement load counts, and release any datasets we
+            # have finished with. This is particularly important
+            # for the shared, default datasets we created above.
+            for data_set in node.inputs:
+                load_counts[data_set] -= 1
+                if load_counts[data_set] < 1 and data_set not in pipeline.inputs():
+                    catalog.release(data_set)
+            for data_set in node.outputs:
+                if load_counts[data_set] < 1 and data_set not in pipeline.outputs():
+                    catalog.release(data_set)
+
+    def run_only_missing(
```

Comment:
1. I implemented this because I felt it was especially relevant for Dask. The `DaskDataSet` actually publishes data to the cluster, so there's less progress lost than in the case of `MemoryDataSet` in case of an error.
2. I did test it months ago when I wrote the behavior, but I probably hacked it in somewhere for testing. Not aware of an easy call off the top of my head (but it probably should be!). Looks like it was raised a few years ago (#30, #60) but decided against back then.

Reply:
Wow, great work searching the archives to find that! I wasn't even aware that `run_only_missing` existed until a couple of months ago when Nikos mentioned it.